Refactor client game code

nuhakala opened the pull request:
This PR does a major refactor to client side game code. Goal is to make code more readable and maintainable, not to change any existing behavior. I have done a fair amount of testing and have not noticed any changes. However, at least when I tried to simplify `getFriction`, it resulted in different behavior and I reverted the change. If you experience any changes, please let me know and I will fix it.

Sorry for huge PR, I considered about splitting it into two but was not sure if that would achieve anything. Following the changes is little complicated due to renaming variables and moving code around, so thought that it might be simplest to just put everything into one PR, especially when the track parsing was removed completely from client. If you want, I will break this into multiple PRs.

Features:
- Creates own class for map and tile and moves the related code from gamecanvas and gamebackgroundcanvas to these classes.
- Deobfuscates (almost) every variable in gamecanvas
- Deobfuscates many variables in hackedshot
- Refactors the client to use track classes from shared library
- Deletes code related to ads in track/map context. Removing it should not affect anything, and I was encouraged to remove it because I saw some ad code removals in commit history.
- Removes some dead code and simplifies few if-clauses.

Other notes:
- Creating gamestate class is little questionable at this point. I created it because I thought that I could refactor the shot-specific functionality to its own class similarly to hackedshot, but that is not so simple so did not do it.
- In my opinion the gamecanvas still holds at least classes `Game`, `Shot`, and some kind of interface for rendering graphics.

Advances issue #69

nuhakala committed:
Refactor: create classes Track and Map

Create new classes Track and Map and extract related game logic
to them.

nuhakala committed:
Refactor: create class Tile and move related logic there

nuhakala committed:
Refactor: Move movable block check into Map class

nuhakala committed:
Refactor: deobfuscate a lot of variables

nuhakala committed:
Client: Reset tile frictions

nuhakala committed:
Client: Add comments and simplify if clause

nuhakala committed:
Client: Rename vars in hackedshot

nuhakala committed:
Refactor: remove client local Track and use shared lib Track

nuhakala committed:
Client: Fix hackedshot variables

nuhakala committed:
Client: Add invisible tile comment back

nuhakala committed:
Client: Remove comments

nuhakala commented:
@StenAL @pehala  you maybe want to also review

StenAL commented on client/src/main/java/agolf/game/Map.java:
```diff
@@ -0,0 +1,323 @@
+     * If input letters are D,E,F,G,H,I, the current tile is exactly the same as an adjacent one so
+     * one is selected, depending on the input letter.
+     */
+    public void parse(String inputData) {
```
It's a bit unfortunate that this logic isn't in the constructor since it means it's possible to have an invalid Map where `parse` hasn't been called and all the fields are uninitialized. Maybe it's possible to only create the `Map` in a track when all the data is available and can be parsed in the constructor?

The same applies for creating `Track`s.

nuhakala replied:
Fixed this now. I needed to remove updating the tiles in map initialization but it did not seem to affect anything.

Did not do anything to `Track`. There the parsing is little more established already and tbh, I am little reluctant to include so big refactor to this PR as well.

StenAL commented on client/src/main/java/agolf/game/GameBackgroundCanvas.java:
```diff
@@ -6,46 +6,30 @@
-        }
+        this.setSize(this.trackWidth, this.trackHeight);
+        this.map = new Map(49, 25);
+        // this.track = new Track(map);
```
Nitpick: commented out code, can be removed

StenAL commented on client/src/main/java/agolf/game/GameBackgroundCanvas.java:
```diff
@@ -60,28 +44,20 @@ public void paint(Graphics g) {
     // this.gameCanvas.createMap(16777216);
     // 16777216 == grass
-    protected void createMap(int var1) {
+    protected void createMap(int tile) {
```
It looks to me that all this function does is initialize a map as all grass-tiles so this function could maybe be renamed for clarity to something like `initializeMap`.

Also, this function is only ever called with `tile` as 16777216, so the parameter could be eliminated and code simplified a bit.

nuhakala replied:
Removed the parameter and renamed to `initializeGameArea`. I thought that `initializeMap` could be misleading as we have `map` variable in the class.

StenAL commented on shared/src/main/java/org/moparforia/shared/tracks/parsers/VersionedTrackFileParser.java:
```diff
@@ -75,7 +76,16 @@ private Track constructTrack(Map<String, Object> parsed) {
+            }
+        } else {
+            // should throw error
+        }
```
These lines can be removed, they do nothing

StenAL commented on shared/src/main/java/org/moparforia/shared/tracks/Track.java:
```diff
@@ -9,12 +9,11 @@
     private String map;
-
     private Set<TrackCategory> categories;
+    private boolean[] specialSettings;
```
Could you document what the special settings are? I see some comments in the `Tile` class but it's pretty hard to find them when looking at the `Track` class here.

nuhakala replied:
Added.

StenAL commented on shared/src/main/java/org/moparforia/shared/tracks/parsers/VersionedTrackFileParser.java:
```diff
@@ -95,15 +105,77 @@ public TrackStats parseStats(Path path) throws IOException {
+        LocalDate bestTime = (LocalDate) parsed.getOrDefault("bestTime", LocalDate.now());
+        String bestPlayer = (String) parsed.getOrDefault("bestPlayer", "");
+        LocalDate lastBestTime = (LocalDate) parsed.getOrDefault("lastBestTime", LocalDate.now());
+        String lastBestPlayer = (String) parsed.getOrDefault("lastBestPlayer", "");
```
It looks like this data is missing for most (all?) tracks which causes the game canvas to say `best result last by Today` for each track. I think this data should only be sent to the client if it actually exists, or the client should do some filtering to not display if it's using the default values.

<img width="260" height="66" alt="image" src="https://github.com/user-attachments/assets/85bd46cf-2b79-4790-952f-09de312fee0b" />

nuhakala replied:
The rendering part actually took this into account already: it made checks whether the last best result is null and if it was, did not render it at all. Hence I changed `getOrDefault` to simply `get` so that the value is null when it does not exist.

StenAL commented on client/src/main/java/agolf/game/Tile.java:
```diff
@@ -0,0 +1,151 @@
+    }
+
+    public Tile(int code) {
+        // this.update(code);
```
Nit: Commented out code, can be removed

StenAL commented on client/src/main/java/agolf/game/GameCanvas.java:
```diff
@@ -700,118 +675,111 @@ public synchronized void keyPressed(KeyEvent event) {
+            this.track = parser.parseTrackFromString(commandLines);
+            this.map.parse(track.getMap());
+            this.trackStats = parser.parseStatsFromString(commandLines);
+            // parseSuccessful = this.track.parse(commandLines);
```
Nit: Commented out code, can be removed.

StenAL commented on client/src/main/java/agolf/game/GameCanvas.java:
```diff
@@ -944,35 +912,35 @@ protected void setPlayerNamesDisplayMode(int mode) {
 
-    protected String method142() {
-        if (this.gameState != 1) {
+    protected String getEncodedCoordinates() {
```
Is this the correct name for the function? It modifies the state which is unusual for a getter.

nuhakala replied:
Changed the name to `encodeCoordinates`, did not come up with a better name and there is already counterpart `decodeCoords` so I  guess it is in line with that.

StenAL reviewed (commented):
Looks mostly good and the changes are really nice since it untangles a lot of the complexity in the codebase!

One issue I have is that some commits fail the build. This makes bisecting issues difficult in the future since this project uses merge commits which preserves your commit history.

It would be great if you could fix that, either by squashing commits or by rebasing and fixing them one-by-one. For reference, I used `git rebase upstream/master -x "mvn clean && mvn install"` to test whether each commit passes the build.

nuhakala committed:
Client: Improve naming + simplifications

nuhakala commented:
@StenAL Thanks for the review and sorry for the slow response. I did `git rebase upstream/master -x "mvn clean && mvn install"` and fixed the issues. I also addressed the other problems, please take a look.

In addition, I reworded the commit messages to better align other commit messages in the history.

StenAL reviewed (approved):
Looks really good, thanks for this!

This is great work since it's deobfuscating the most complex part of the whole application. Everything looks way more understandable and decoupled than before.

PhilippvK commented:
Thank you for the awesome contribution. I really appreciate the efforts.
